---
title: "Visualising NGS data with ggplot and ggbio"
author: "Mark Dunning"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
output: html_document
---

## More-advanced graphics in R

- Base graphics in R use a canvas model
    + series of instructions that sequentially fill the plotting canvas
    
- ggplot2 employs a `grammar of graphics approach`
- The components are
    + a dataset
    + geometric object that is visual representation of the data
        + e.g. points, lines, etc
    + mapping of variables to visual properties of plot
        + ***aesthetics***
    + (statistical summarisation rule)
    + (coordinate system)
    + (facet specification)

## ggplot2 overview
ggplot2 [cheat-sheet](https://www.rstudio.com/wp-content/uploads/2015/03/ggplot2-cheatsheet.pdf)

![cheatsheet](images/ggplot2-csheat.png)

## Plot Comparison

```{r fig.height=3}
x <- 1:10
y <- 2*x
plot(x,y)
```

```{r fig.height=3}
library(ggplot2)
df <-data.frame(x,y)
ggplot(df, aes(x=x,y=y)) + geom_point()

```

## Plot construction
- ggplot2 needs data as a data frame
- It needs to be ***long*** format

```{r}
df <- data.frame(A = rnorm(5,3), B=rnorm(5,1))
df
```

The `tidyr` package is the last in a long line of packages (reshape, reshape2,..) that will manipulate data into a more covenient form for analysis. 

```{r}
library(tidyr)
df2 <- gather(df)
df2

```


## Plot construction



```{r}
ggplot(df2, aes(x = variable,y=value)) + geom_boxplot()
```


## Plot construction

```{r fig.height=3}
ggplot(df2, aes(x = variable,y=value,fill=variable)) + geom_boxplot()
```

## Updating a plot

- ggplot2 will easily re-drawn a plot as new variables are added
    + a real advantage!
```{r fig.height=3}
df <- data.frame(A = rnorm(5,3), B=rnorm(5,1),C=rnorm(5,2))
df2 <- melt(df)
ggplot(df2, aes(x = variable,y=value,fill=variable)) + geom_boxplot()
```


# ggplot2 Exercise

## Plotting the annotation table



    
## Introducing ggbio

- A consistent representation of ranges and genomic data helps with visualisation
- The `ggbio` package is a toolkit for producing publication-quality images from genomic data
- It extends the ***Grammar of Graphics*** approach taken by `ggplot2`
- It knows about the standard Bioconductor classes we have already introduced
- Published in [Genome Biology](http://www.genomebiology.com/2012/13/8/R77)
![ggbio](images/ggbio.png)

## The autoplot function

- Guesses what type of plot you want from the data
- Figures out the x and y coordinates
```{r message=FALSE}
library(ggbio)
library(GenomicAlignments)
mygene <- GRanges("17", ranges=IRanges(7577851, 7598063))
mygene.reads <- readGAlignments(file="../alignedReads/paired.bam",param=ScanBamParam(which=mygene),use.names = TRUE)
```

## Can choose a summary statistic

```{r message=FALSE}
autoplot(mygene.reads,stat="coverage")
```

## Plotting gene structure

```{r message=FALSE}
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
exons <- exonsBy(txdb,"gene")
autoplot(txdb,which=exons[["7157"]])
```

## Combining plots

- plots made by `ggplot2` cannot be customised in the usual way with `par`
    + e.g. `par(mfrow=c(1,2))`
- the `tracks` function in `ggbio` can do this job    
- x-axis structure is consistent between plots

```{r message=FALSE}
tracks(autoplot(txdb,which=exons[["7157"]]),
autoplot(mygene.reads,stat="coverage"))
```

## Different layouts available

- Can easily switch between different plot layouts
    + geoms in ggplot2 terms
- Also set aesthetics from properties in the data
    + using `aes`; like in `ggplot2`
- e.g. Manhattan plot

![manhat](images/gb-2012-13-8-r77-3.jpg)

## Karyogram

Karyogram

![kgram](images/gb-2012-13-8-r77-4.jpg)

## Circular

![circos](images/gb-2012-13-8-r77-5.jpg)

## Practical time

- Use `ggplot2` and `ggbio` to explore the RNA-seq results
- Feel free to experiment
- Or check-out the vignettes for either package