---
title: "Visualising NGS data with R and Bioconductor"
author: "Mark Dunning"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
output: html_document
---

# Graphics in R

The R language has extensive graphical capabilities.

Graphics in R may be created by many different methods including base graphics and more advanced plotting packages such as lattice.

<div id="buildingblocks" align="center">
  <img src="images/plotsinR.jpg" alt="">
</div>  

The ggplot2 package was created by Hadley Wickham and provides a intuitive plotting system to rapidly generate publication quality graphics.

ggplot2 builds on the concept of the "Grammar of Graphics" (Wilkinson 2005, Bertin 1983) which describes a consistent syntax for the construction of a wide range of complex graphics by a concise description of their components.


# Why use ggplot2

The structured syntax and high level of abstraction used by ggplot2 should allow for the user to concentrate on the visualisations instead of creating the underlying code.

On top of this central philosophy ggplot2 has:

- Increased flexible over many plotting systems.
- An advanced theme system for professional/publication level graphics.
- Large developer base -- Many libraries extending its flexibility.
- Large user base -- Great documentation and active mailing list.


# Grammar of Graphics

## How to build a plot from its components.

<div id="buildingblocks" align="center">
  <img src="presentation1/Slide1.jpg" alt="">
</div>  


## How ggplot2 builds a plot.

<div id="buildingblocks" align="center">
  <img src="presentation1/Slide2.jpg" alt="">
</div>  

Overview of example code for the ggplot2 scatter plot.

```{r complex,eval=F}

ggplot(data = <default data set>, 
       aes(x = <default x axis variable>,
           y = <default y axis variable>,
           ... <other default aesthetic mappings>),
       ... <other plot defaults>) +

       geom_scatter(aes(size = <size variable for this geom>, 
                      ... <other aesthetic mappings>),
                  data = <data for this point geom>,
                  stat = <statistic string or function>,
                  position = <position string or function>,
                  color = <"fixed color specification">,
                  <other arguments, possibly passed to the _stat_ function) +

  scale_<aesthetic>_<type>(name = <"scale label">,
                     breaks = <where to put tick marks>,
                     labels = <labels for tick marks>,
                     ... <other options for the scale>) +
  
  ggtitle("Graphics/Plot")+
  xlab("Weight")+
  ylab("Height")+

  theme(plot.title = element_text(colour = "gray"),
        ... <other theme elements>)
    
```

## What users are required to specify in ggplot2 to build a plot.

<div id="buildingblocks" align="center">
  <img src="presentation1/Slide3.jpg" alt="">
</div>  


# Getting started with ggplot2



```{r load_packages, echo=FALSE, eval=TRUE,warning=F,message=F}
suppressPackageStartupMessages(library(tidyr))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(stringr))
suppressPackageStartupMessages(library(lubridate))
```


```{r present_clean}
library(ggplot2)
library(dplyr)

anno <- read.csv("../annotate-and-filter/myanno.hg19_multianno.joined.csv")
anno <- tbl_df(anno)
anno
```

## Our first ggplot2 graph

As seen above, in order to produce a ggplot2 graph we need a minimum of:-

- Data to be used in graph
- Mappings of data to the graph (aesthetic mapping)
- What type of graph we want to use (The geom to use).

In the code below we define the data as our cleaned patients data frame.

```{r ggplot_only}
pcPlot <- ggplot(data=anno)
class(pcPlot)
pcPlot$data[1:4,]
```

Now we can see that we have gg/ggplot object (pcPlot)  and in this the data has been defined.

Important information on how to map the data to the visual properties (aesthetics) of the plot as well as what type of plot to use (geom) have however yet to specified. 

```{r missing_rest}
pcPlot$mapping
pcPlot$theme
pcPlot$layers
```


The information to map the data to the plot can be added now using the aes() function.
```{r ggplot_aes}
pcPlot <- ggplot(data=anno)

pcPlot <- pcPlot+aes(x=X1000g2014oct_eur,y=X1000g2014oct_afr)

pcPlot$mapping
pcPlot$theme
pcPlot$layers
```

But we are still missing the final component of our plot, the type of plot to use (geom).

Below the geom_point function is used to specify a point plot, a scatter plot of Height values on the x-axis versus Weight values on the y values.

```{r ggplot_aes_geom}
pcPlot <- ggplot(data=anno)

pcPlot <- pcPlot+aes(x=X1000g2014oct_eur,y=X1000g2014oct_afr)
pcPlot <- pcPlot+geom_point()

pcPlot$mapping
pcPlot$theme
pcPlot$layers

```

Now we have all the components of our plot, we need we can display the results.

```{r ggplot_aes_geom_display}
pcPlot

```


More typically, the data and aesthetics are defined within ggplot function and geoms applied afterwards.

```{r ggplot_simple_geom_point}
pcPlot <- ggplot(data=anno,
                 mapping=aes(x=X1000g2014oct_eur,y=X1000g2014oct_afr))
pcPlot+geom_point()
```


# Geoms - Plot types

As we have seen, an important element of a ggplot is the geom used. Following the specification of data, the geom describes the type of plot used.

Several geoms are available in ggplot2:-

* geom_point() - Scatter plots
* geom_line() - Line plots
* geom_smooth() - Fitted line plots
* geom_bar() - Bar plots
* geom_boxplot() - Boxplots
* geom_jitter() - Jitter to plots
* geom_hist() - Histogram plots
* geom_density() - Density plots
* geom_text() - Text to plots
* geom_errorbar() - Errorbars to plots
* geom_violin() - Violin plots

## Geoms - Line plots



```{r, line_simple}
pcPlot <- ggplot(data=anno,
        mapping=aes(x=X1000g2014oct_eur,y=X1000g2014oct_afr))

pcPlot_line <- pcPlot+geom_line() 

pcPlot_line

```

```{r, smooth_simple}
pcPlot <- ggplot(data=anno,
        mapping=aes(x=X1000g2014oct_eur,y=X1000g2014oct_afr))

pcPlot_smooth <- pcPlot+geom_smooth() 

pcPlot_smooth

```


## Geoms - Bar and frequency plots

```{r, bar_simple}
pcPlot <- ggplot(data=anno,
        mapping=aes(x=Func.refGene))

pcPlot_bar <- pcPlot+geom_bar() 

pcPlot_bar


```


```{r}
pcPlot_bar + coord_flip() 
```


```{r, histogram_simple}
pcPlot <- ggplot(data=anno,
        mapping=aes(x=X1000g2014oct_eur))

pcPlot_hist <- pcPlot+geom_histogram() 

pcPlot_hist

```


```{r, density_simple}
pcPlot <- ggplot(data=anno,
        mapping=aes(x=X1000g2014oct_eur))

pcPlot_density <- pcPlot+geom_density() 

pcPlot_density

```


## Geoms - Box and violin plots

```{r, boxplot_simple}

pcPlot <- ggplot(data=anno,
        mapping=aes(x=Func.refGene,y=X1000g2014oct_eur))

pcPlot_boxplot <- pcPlot+geom_boxplot() 

pcPlot_boxplot
```

```{r, violin_simple}

pcPlot <- ggplot(data=anno,
        mapping=aes(x=Func.refGene,y=X1000g2014oct_eur))

pcPlot_violin <- pcPlot+geom_violin() 

pcPlot_violin
```

An overview of geoms and thier arguments can be found at ggplot2 documentation or within the ggplot2 cheatsheet.

-[ggplot2 documentation](http://docs.ggplot2.org/current/)

-[ggplot2 Cheatsheet](http://sape.inf.usi.ch/quick-reference/ggplot2/geom)

# Aesthetics

In order to change the property on an aesthetic of a plot into a constant value (e.g. set colour of all points to red) we can supply the colour argument to the geom_point() function.

```{r, scatter_coloured}
pcPlot <- ggplot(data=anno,
                 mapping=aes(x=X1000g2014oct_eur,y=X1000g2014oct_afr))
pcPlot+geom_point(colour="red")
```


As we discussed earlier however, ggplot2 makes use of aesthetic mappings to assign variables in the data to the properties/aesthetics of the plot. This allows the properties of the plot to reflect variables in the data dynamically.

In these examples we supply additional information to the aes() function to define what information to display and how it is represented in the plot.

First we can recreate the plot we saw earlier.

```{r, scatter_simple}
pcPlot <- ggplot(data=anno,
                 mapping=aes(x=X1000g2014oct_eur,y=X1000g2014oct_afr))
pcPlot+geom_point()
```

Now we can adjust the aes mapping by supplying an argument to the colour parameter in the aes function. (Note that ggplot2 accepts "color" or "colour" as parameter name)


This simple adjustment allows for identifaction of the separation between male and female measurements for height and weight.

```{r, scatter_aes_sexColour}
pcPlot <- ggplot(data=anno,
                 mapping=aes(x=X1000g2014oct_eur,y=X1000g2014oct_afr,colour=GT.TUMOUR))
pcPlot+geom_point()
```

Similarly the shape of points may be adjusted.

```{r, scatter_aes_sexShape}
pcPlot <- ggplot(data=anno,
                 mapping=aes(x=X1000g2014oct_eur,y=X1000g2014oct_afr,shape=GT.TUMOUR))
pcPlot+geom_point()
```

The aesthetic mappings may be set directly in the `geom_point()` function as previously when specifying red. This can allow the same ggplot object to be used by different aesethetic mappings and varying geoms

```{r, aes_in_geom}
pcPlot <- ggplot(data=anno)
pcPlot+geom_point(aes(x=X1000g2014oct_eur,y=X1000g2014oct_afr,colour=GT.TUMOUR))
pcPlot+geom_point(aes(x=X1000g2014oct_eur,y=X1000g2014oct_afr,colour=GT.NORMAL))
pcPlot+geom_point(aes(x=X1000g2014oct_eur,y=X1000g2014oct_afr,colour=GT.TUMOUR,shape=GT.TUMOUR,size=DP))
pcPlot+geom_violin(aes(x=is.na(GT.TUMOUR),y=X1000g2014oct_eur,fill=GT.TUMOUR))
```


Again, for a comprehensive list of parameters and aesthetic mappings used in geom_*type*  functions see the ggplot2 documentation for individual geoms by using ?geom_*type*

```{r, helpforArguments}
?geom_point
```

or visit the ggplot2 documentations pages and cheatsheet

- [ggplot2 documentation](http://docs.ggplot2.org/current/)

- [Cheatsheet](http://sape.inf.usi.ch/quick-reference/ggplot2/geom)


# Facets

One very useful feature of ggplot is faceting. 
This allows you to produce plots subset by variables in your data.

`ggplot2` is not compatible with the way that plots are constructed in *base* R, so we cannot use techniques like `mfrow=c(..)` and `mfrow=c(..)` to split our plotting window into different cells. Fortunately, facet provides a more-intuitive way of composing a plot based on the variables in the data.

To facet our data into multiple plots we can use the *facet_wrap* or *facet_grid* function specifying the variable we split by.

The *facet_grid* function is well suited to splitting the data by two factors.


To split by one factor we can apply the facet_grid() function ommiting the variable before the "~"" to facet along columns in plot.

<div align="center">
facet_grid(~Columns)
</div> 
```{r, facet_grid_BySex}
pcPlot <- ggplot(data=anno,aes(x=X1000g2014oct_eur,y=X1000g2014oct_afr))+geom_point()
pcPlot + facet_grid(~GT.TUMOUR)
```

To split along rows in plot, the variable is placed before the "~.".

<div align="center">
facet_grid(Rows~.)
</div> 

```{r, facet_grid_SexBy}
pcPlot <- ggplot(data=anno,aes(x=X1000g2014oct_eur,y=X1000g2014oct_afr))+geom_point()
pcPlot + facet_grid(GT.TUMOUR~.)
```

The *facet_wrap()* function offers a less grid based structure but is well suited to faceting data by one variable.

For *facet_wrap()* we follow as similar syntax to *facet_grid()*

```{r, facet_Wrap_BySmokes}
pcPlot <- ggplot(data=anno,aes(x=X1000g2014oct_eur,y=X1000g2014oct_afr))+geom_point()
pcPlot + facet_wrap(~Chr)
```

A point to note here is that the order of the chromosomes is *alphabetical* rather than numeric. If we wanted a more natural order, we can *re-factor* the variable in the data frame to get the desired effect.

```{r}
anno <- mutate(anno, Chr=factor(Chr, levels=c(1:22, "X","Y")))
```

```{r, facet_Wrap_BySmokes2}
pcPlot <- ggplot(data=anno,aes(x=X1000g2014oct_eur,y=X1000g2014oct_afr))+geom_point()
pcPlot + facet_wrap(~Chr)
```


******
### Exercise

- The number of variants observed on each Chromosome

```{r echo=FALSE}
ggplot(anno, aes(x=Chr)) + geom_bar()
```

- The distribution of 1000 genomes (european) frequencies for each chromosome

```{r echo=FALSE}
ggplot(anno, aes(x = Chr,y=X1000g2014oct_eur)) + geom_boxplot()
```

- The distribution of 1000 genomes (european) frequencies depending on whether the variant has a dbSNP ID or not

```{r echo=FALSE}
mutate(anno, IndbSNP = !is.na(snp129)) %>% 
ggplot(aes(x = IndbSNP,y=X1000g2014oct_eur)) + geom_boxplot()
```


- The numbers of exonic, intergenic and intronic (ONLY) variants for each chromosome
    + You'll need to `filter` the data frame appropriately

```{r echo=FALSE}
filter(anno,Func.refGene %in% c("exonic", "intergenic","intronic","")) %>% 
  ggplot(aes(x = Func.refGene)) + geom_bar() + facet_wrap(~Chr)  
```

- The depth observed at each variant (`DP`) along each chromosome

```{r echo=FALSE}
 ggplot(anno, aes(x = Start, y= DP)) + geom_point() + facet_wrap(~Chr)
```

******

In the last plot we started to look at our data on a genome-wide scale. 

```{r echo=FALSE}
path.to.bam <- "/media/dunnin01/Backup/summer-school/input"
```

    
## Introducing ggbio

- A consistent representation of ranges and genomic data helps with visualisation
- The `ggbio` package is a toolkit for producing publication-quality images from genomic data
- It extends the ***Grammar of Graphics*** approach taken by `ggplot2`
- It knows about the standard Bioconductor classes we have already introduced
- Published in [Genome Biology](http://www.genomebiology.com/2012/13/8/R77)

So the first stage is to transform our data into a `GenomicRanges` object

```{r}
library(GenomicAlignments)
anno.gr <- GRanges(paste0("chr",anno$Chr), IRanges(anno$Start,anno$End))
mcols(anno.gr) <- select(anno, -(Chr:End))
anno.gr
```

`ggbio` has a convenient function called `autoplot` which will try and predict what visualisation we want from a particular data type.

In this case, we can plot the coordinates of a set of variants. You should recognise from the distinctive grey colour scheme that `ggbio` is using `ggplot2` to create the plot.


```{r}
library(ggbio)
autoplot(anno.gr[1:10])
```


To visualise our results on a whole-genome scale, we can us the `plotGrandLinear` function. This function is intended to produce *"Manhattan plots"* commonly used to present whole-genome results (e.g. results from a GWAS study). We need to specify what is to be plotted on the y-axis. Typically this would be some form of significance, for example  -log(p-value). `ggbio` will figure out where all the points need to be plotted on the x-axis

```{r}
library(ggbio)

plotGrandLinear(anno.gr, aes(y = DP))
```

As `ggbio` is writen in the same style as `ggplot2`, we can set other aesthetics such as colour to be variable found in the `mcols` section of the object we are plotting

```{r}
plotGrandLinear(anno.gr, aes(y = DP,col=GT.TUMOUR),legend = TRUE)
```

We can also introduce "facetting" into the mix:-

```{r}
plotGrandLinear(anno.gr, aes(y = DP,col=GT.TUMOUR),legend = TRUE) + facet_wrap(~GT.TUMOUR)
```

There are also a number of facilities available fo looking at individuals variants and genomic regions of interest.

We see the first variant is located in the gene "WASH7P". The `ggbio` package will allow us to visualise where this variant is located with respect to the gene. Another type of object that `ggbio` can recognise is the `TxDb....` databases of gene models that we introduced previously. Here we also load a `genesymbol` object, which is a convenient way of getting ranges for genes based on symbol.

```{r message=FALSE}
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
data(genesymbol, package="biovizBase")
wh <- genesymbol["WASH7P"]
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene

autoplot(txdb,which=wh)
```


We can combine plots using the `tracks` function. This will automatically align features on the x-axis (genome) for us.

```{r message=FALSE}
tracks(autoplot(txdb,which=wh), 
       autoplot(anno.gr[2])
)

```

`ggbio` is also able to incorporate our aligned reads. We can point the `autoplot` function directly to the location of the bam file and it will generate coverage track automatically.

```{r message=FALSE}
seqlevelsStyle(wh) <- "Ensembl"
wh <- keepSeqlevels(wh,"1")

bam <- paste0(path.to.bam, "/HCC1143.mix1.n40t60.bam")

autoplot(bam, which=wh)
```

If we want to visualise the actual reads themselves, we can first import them using the `readGAlignments` function and call the `autoplot` function on the resulting object.

```{r message=FALSE}
reads <- readGAlignments(bam, param=ScanBamParam(which=wh))
autoplot(reads, which=wh)
```


```{r message=FALSE}
library(BSgenome.Hsapiens.UCSC.hg19)

bg <- BSgenome.Hsapiens.UCSC.hg19

autoplot(bam,bsgenome=bg, stat="mismatch",which=wh)
```


```{r message=FALSE}
seqlevelsStyle(txdb) <- "Ensembl"
tracks(Genes=autoplot(txdb,which=wh), 
       Variant=autoplot(anno.gr[2]),
       Mismatch=autoplot(bam,bsgenome=bg, stat="mismatch",which=wh)
)
```

```{r message=FALSE}
seqlevelsStyle(txdb) <- "Ensembl"
tracks(Genes=autoplot(txdb,which=wh), 
       Variant=autoplot(anno.gr[2]),
       Mismatch=autoplot(bam,bsgenome=bg, stat="mismatch",which=wh),
       heights=c(4,1,4)
)
```


******
### Exercise

The *cansar* browser notes a deletion in the region `Chr1:152555707-152586541` for [HCC1143](https://cansar.icr.ac.uk/cansar/cell-lines/HCC1143/copy_number_variation/chromosome_1/)

- Do we see this in our data? 
```{r echo=FALSE,message=FALSE}
wh <- GRanges("1", IRanges(152551500,152589096))
bam <- paste0(path.to.bam, "/HCC1143.mix1.n40t60.bam")
#reads <- readGAlignments(file = bam, param=ScanBamParam(which=wh),use.names = TRUE)
autoplot(bam, which=wh)
reads <- readGAlignments(bam, param=ScanBamParam(which=wh))
seqlevelsStyle(txdb) <- "Ensembl"
tracks(Reads=autoplot(reads),
       Genes = autoplot(txdb, which=wh),
       heights=c(4,1)
)


```


******

The `Gviz` package was finer-control over the plotting of data as tracks.

```{r}
library(Gviz)

vars <- AnnotationTrack(anno.gr[1:10])
gtrack <- GenomeAxisTrack()
itrack <- IdeogramTrack(genome = "hg19", chromosome = "chr1")

library(TxDb.Hsapiens.UCSC.hg19.knownGene)

txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
grtrack <- GeneRegionTrack(txdb)

plotTracks(list(itrack, gtrack, vars, grtrack),from=10000,to=20000,collapseTranscripts=TRUE,transcriptAnnotation="symbol")

varpos <- start(anno.gr[2])

```

```{r}
reads <- readGAlignments(bam, use.names=TRUE,param=ScanBamParam(which = GRanges("1",IRanges(varpos-20, varpos+20)),what=c("cigar","mapq","seq","flag")))
seqlevelsStyle(reads) <- "UCSC"

altrack <- AlignmentsTrack(start=start(reads),end=end(reads),chromosome = seqnames(reads),cigar=mcols(reads)$cigar,genome="hg19")

strack <- SequenceTrack(Hsapiens, chromosome = "chr1")
plotTracks(list(itrack, vars, altrack), from = varpos-20, to = varpos+20, cex = 0.8,collapseTranscripts=TRUE)

```

```{r}
ht <- HighlightTrack(trackList = list(altrack, strack), start=varpos-1,width=2,chromosome="chr1")
```

```{r}
plotTracks(list(itrack,vars,ht,grtrack), from = varpos-20, to = varpos+20, cex = 0.8)

```


